core/solr_index.py

```diff
@@ -573,20 +571,18 @@ def index_titles(since=None):
         if row == None:
             break
         title = models.Title.objects.get(lccn=row[0])
-        index_title(title, solr)
+        index_title(title)
         count += 1
         if count % 100 == 0:
             _log.info("indexed %s titles" % count)
             reset_queries()
-            solr.commit()
-    solr.commit()
+            conn().commit()
+    conn().commit()
```

Comment:
Short answer: what's happening here is fine and needs to get implemented before we go refactoring ONI.

Lone answer:

First, I'm not sure there's a whole lot of benefit in sharing a solr connection.  Solr requests are just HTTP requests, so sharing them requires HTTP 2 on both sides, which may not even be an option.  I think Solr can talk HTTP 2, but it may require special configuration, I have no idea.  I don't know if pysolr even has an HTTP 2 option.

Aside from that, it wasn't being shared when we had solrpy.  I wouldn't go changing something else just to change it unless there's a very good reason.  The update is in place and currently I have confidence that it's doing exactly what the old system did, just with a new library.

But even if this were a quick and safe adjustment to the code, and fully supported in pysolr and Solr, the benefits are still going to be very minimal.  The batch ingest is already sharing a Solr connection (this wasn't changed, it was doing this with solrpy as well), so there's nothing to be gained there.  The front-end searches are scattered and can hit any one of multiple ONI instances (depending on the WSGI thread setup), and the whole ONI application is restarted regularly (depending on the WSGI maximum request count).  It's hard to imagine seeing any improvement in performance from such a change.

Reply:
I wasn't thinking to refactor all of ONI to re-use a single object, but at least restore these lines you dropped where the connection object is stored in a variable: https://github.com/open-oni/open-oni/pull/561/commits/49966f11fb569177e10f78eec26b53e3ab379ee1#diff-65d554ce9c974916d220c73ee1aa6de6L563

Functions where there isn't more than one `conn()` probably don't need it though. Don't know if this is micro-optimizing territory, but it was handled that way before with solrpy. Not 100% on whether this actually saves some HTTP requests, but it probably cuts down at least a little on object reconstruction in Python.

If there is a way to share a connection to Solr app-wide, that'd definitely be a separate issue and PR :+1:

Reply:
Missed the context of this one entirely.  UGH.  Sorry, this one's 100% my fault.